torch/distributed/_tensor/_dispatch.py

```diff
@@ -438,21 +446,27 @@ def _try_replicate_dtensor_spec_in_missing_dim(
         mesh: "DeviceMesh",
     ) -> DTensorSpec:
         # util function to produce a new spec for a DTensor arg/kwarg
-        # that puts Replicate() placement in the missing dimension for foreach ops
+        # that puts Replicate() placement in the missing dimension for foreach ops in 2D scenario
+        cur_mesh = dtensor_arg.device_mesh
+        error_msg = (
+            f"{op_call}: DTensor does not support cross-mesh operation yet! "
+            f"Got meshes: {mesh} {cur_mesh}"
+        )
+        # quick throw if implicit replication is not allowed
+        if not self._allow_implicit_replication:
+            raise NotImplementedError(error_msg)
+
         from torch.distributed.device_mesh import _mesh_resources
 
-        cur_mesh = dtensor_arg.device_mesh
-        root_mesh = _mesh_resources.get_root_mesh(cur_mesh)
-        if (
-            self._allow_implicit_replication
-            and "foreach" in op_call.__name__
-            and root_mesh == mesh
-        ):
-            placements = [Replicate() for _ in range(root_mesh.ndim)]
+        # only allow implicit replication if cur_mesh and mesh have the same root mesh.
+        cur_mesh_root = _mesh_resources.get_root_mesh(cur_mesh)
+        mesh_root = _mesh_resources.get_root_mesh(mesh)
+        if "foreach" in op_call.__name__ and cur_mesh_root == mesh_root:
```

Comment:
Why restricting this to for_each ops only?

IMO if user turn on implicit replication flag, every op who have one arg that is not a DTensor should be implicitly replicated here